### PR TITLE
Add portable DebugMode log serialization

### LINKED ISSUE #168973: DebugMode sample code is unrealistic for real world training use case
### 🐛 Describe the bug

The sample code for DebugMode looks like this:

```
            # Run model first time
            with DebugMode() as debug_mode, DebugMode.log_tensor_hashes():
                model(x)
                logs1 = debug_mode.logs

            # Run again, in exactly the same way
            with DebugMode() as debug_mode, DebugMode.log_tensor_hashes():
                model(x)
                logs2 = debug_mode.logs

            mismatches = DebugMode.check_hash_mismatches(logs1, logs2)
            for m in mismatches:
                print(f"{m['call']}: hash diff {m['rel_diff']:.2e}")
```

But this is unrealistic in a real world training setting, where it is likely that I am trying to diff two different model configs which can't even run in the same process at the same time.

cc @chauhang @penguinwu @yushangdi @pianpwk 

### Versions

main

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185010

DebugMode hash mismatch debugging previously only worked with in-memory log objects. The public example compared two DebugMode runs in one process, but real training comparisons often need to run different model configs in separate processes. The raw log records keep runtime operator/function objects, so saving them directly with torch.save is not reliable and can fail on non-pickleable objects.

Add JSON-backed DebugMode.save_logs() and DebugMode.load_logs() helpers that store the call structure, rendered arguments, stack traces, and hash log payloads needed by check_hash_mismatches() without serializing runtime operator objects. The portable format intentionally omits .record payloads because record_outputs() can contain tensors that are not needed for mismatch comparison. Non-finite float hash values are encoded explicitly so files remain valid JSON.

The check_hash_mismatches() usage example now shows two independent model-config runs saving logs and comparing the loaded files. Tests cover save/load mismatch comparison, interaction with record_outputs(), and non-finite hash round-tripping.

Fixes #168973

Generated by my agent

Test Plan:

python test/distributed/tensor/debug/test_debug_mode.py -k TestDebugModeLogSerialization

python test/distributed/tensor/debug/test_debug_mode.py TestDTensorDebugMode.test_check_hash_mismatches TestDTensorDebugMode.test_check_structure_mismatches

lintrunner -a torch/utils/_debug_mode/_calls.py torch/utils/_debug_mode/_mode.py test/distributed/tensor/debug/test_debug_mode.py

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the Claude review feedback by documenting why _RedistributeCall deserialization restores is_outer_call after construction, and by adding empty-log plus redistribute outer-call serialization coverage. I also rebased on current main and added the release notes: distributed (dtensor) label.

**Review by pianpwk — APPROVED** [review]
thanks!!

**Comment by jansel** [comment]
Agent, share an example of what the new serialized output looks like for the case in the original issue.

**Comment by jansel** [comment]
My agent says here is a compact example for the original two-run workflow. A `DebugMode.save_logs(debug_mode.logs, "run1.json")` file for `x.sin().sum()` looks like this structurally, with hash values shortened here:

```json
{
  "format": "torch.utils._debug_mode.DebugMode.logs",
  "version": 1,
  "logs": [
    {
      "type": "op",
      "op_name": "aten::sin",
      "call_depth": 0,
      "args_str": "Tensor(shape=(4, 4), dtype=torch.float32, device=cpu)",
      "kwargs_str": "",
      "output_str": "Tensor(shape=(4, 4), dtype=torch.float32, device=cpu)",
      "log": {
        "type": "dict",
        "items": [
          ["hash", {"type": "scalar", "value": 0.123456}],
          ["input_hash", {"type": "tuple", "items": [{"type": "scalar", "value": 0.654321}]}]
        ]
      },
      "stack_trace": null,
      "fwd_stack_trace": null
    },
    {
      "type": "op",
      "op_name": "aten::sum",
      "call_depth": 0,
      "args_str": "Tensor(shape=(4, 4), dtype=torch.float32, device=cpu)",
      "kwargs_str": "",
      "output_str": "Tensor(shape=(), dtype=torch.float32, device=cpu)",
      "log": {
        "type": "dict",
        "items": [
          ["hash", {"type": "scalar", "value": 0.234567}],
          ["input_hash", {"type": "tuple", "items": [{"type": "scalar", "value": 0.123456}]}]
        ]
      },
      "stack_trace": null,
      "fwd_stack_trace": null
    }
  ]
}
```

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by jansel** [comment]
My agent says I inspected the two current red CI jobs. The DebugMode test suite added by this PR passed in the distributed job, and the remaining failures are in unrelated dtensor_ops / Python 3.14 crossref coverage, so I reran the failed jobs to clear likely unrelated CI failures.
